nixos/modules/services/web-apps/audiobookshelf.nix

```diff
@@ -0,0 +1,113 @@
+{ pkgs
+, lib
+, config
+, ...
+}:
+
+with lib; let
+  cfg = config.services.audiobookshelf;
+in
+
+{
+  options.services.audiobookshelf = {
+    enable = mkEnableOption (lib.mdDoc  "Audiobookshelf, self-hosted audiobook and podcast server");
+
+    host = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = lib.mdDoc ''
+        The host address that Audiobookshelf will listen on.
+      '';
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 8000;
+      description = lib.mdDoc ''
+        The port that Audiobookshelf will listen on.
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = "audiobookshelf";
+      description = lib.mdDoc ''
+        User account under which Audiobookshelf runs.
+      '';
+    };
+
+    group = mkOption {
+      type = types.str;
+      default = "audiobookshelf";
+      description = lib.mdDoc ''
+        Group under which Audiobookshelf runs.
+      '';
+    };
+
+    openFirewall = mkOption {
+      type = types.bool;
+      default = false;
+      description = lib.mdDoc ''
+        Whether to open the firewall for the port in {option}`services.audiobookshelf.port`.
+      '';
+    };
+
+    configDir = mkOption {
+      type = types.str;
+      default = "/etc/audiobookshelf";
+      description = lib.mdDoc ''
+        Configuration directory Audiobookshelf will use.
+      '';
+    };
+
+    metadataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/audiobookshelf";
+      description = lib.mdDoc ''
+        Metadata directory Audiobookshelf will use.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.port ];
+
+    users.groups = mkIf (cfg.group == "audiobookshelf") {
+      audiobookshelf = { };
+    };
+
+    users.users = mkIf (cfg.user == "audiobookshelf") {
+      audiobookshelf = {
+        group = cfg.group;
+        description = "Audiobookshelf Daemon user";
+        isSystemUser = true;
+      };
+    };
```

Comment:
So then should I remove the user & group options?

Reply:
If you decide to not use `DynamicUser`, then you can consider adding the following snippet as well:

```
systemd.tmpfiles.rules = [
  "d '/var/lib/audiobookshelf' 0750 ${cfg.user} ${cgf.group} - -"
  "z '/var/lib/audiobookshelf' 0750 ${cfg.user} ${cgf.group} - -"
];
```

This will make sure that the `/var/lib/audiobookshelf` folder exists else the service will fail to start.

Reply:
This isn't necessary, if you just use `StateDirectory=audiobookshelf`.

Reply:
> This isn't necessary, if you just use `StateDirectory=audiobookshelf`.

In my experience if the directory doesn't automatically exist the service fails.

Reply:
Which is why [`StateDirectory` creates it before running the service](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#RuntimeDirectory=).

Reply:
That only happens in practice if the service is running as root. In this case because the service is running as the `audiobookshelf` user it doesn't work.